pkg/plugins/s3/destination/destination.go

```diff
@@ -18,96 +18,111 @@ import (
+	d.Buffer = make([]sdk.Record, 0, d.Config.BufferSize)
+	d.AckFuncCache = make([]sdk.AckFunc, 0, d.Config.BufferSize)
 
 	// initializing the writer
 	writer, err := writer.NewS3(ctx, &writer.S3Config{
-		AccessKeyID:     s.Config.AWSAccessKeyID,
-		SecretAccessKey: s.Config.AWSSecretAccessKey,
-		Region:          s.Config.AWSRegion,
-		Bucket:          s.Config.AWSBucket,
-		KeyPrefix:       s.Config.Prefix,
+		AccessKeyID:     d.Config.AWSAccessKeyID,
+		SecretAccessKey: d.Config.AWSSecretAccessKey,
+		Region:          d.Config.AWSRegion,
+		Bucket:          d.Config.AWSBucket,
+		KeyPrefix:       d.Config.Prefix,
 	})
 
 	if err != nil {
 		return err
 	}
 
-	s.Writer = writer
-
+	d.Writer = writer
 	return nil
 }
 
-// Teardown gracefully disconnects the client
-func (s *Destination) Teardown() error {
-	return nil // TODO
-}
-
-// Validate takes config and returns an error if some values are missing or
-// incorrect.
-func (s *Destination) Validate(cfg plugins.Config) error {
-	_, err := Parse(cfg.Settings)
-	return err
-}
-
-// Write writes a record into a Destination. Typically Destination maintains an in-memory
+// WriteAsync writes a record into a Destination. Typically Destination maintains an in-memory
 // buffer and doesn't actually perform a write until the buffer has enough
 // records in it. This is done for performance reasons.
-func (s *Destination) Write(ctx context.Context, r record.Record) (record.Position, error) {
-	// If either Destination or Writer have encountered an error, there's no point in
+func (d *Destination) WriteAsync(ctx context.Context, r sdk.Record, ackFunc sdk.AckFunc) error {
+	// If either Destination or Writer have encountered an error, there'd no point in
 	// accepting more records. We better signal the error up the stack and force
 	// the server to maybe re-instantiate plugin or do something else about it.
-	if s.Error != nil {
-		return nil, s.Error
+	if d.Error != nil {
+		return d.Error
 	}
 
-	s.Mutex.Lock()
-	defer s.Mutex.Unlock()
+	d.Mutex.Lock()
+	defer d.Mutex.Unlock()
 
-	s.Buffer = append(s.Buffer, r)
+	d.Buffer = append(d.Buffer, r)
+	d.AckFuncCache = append(d.AckFuncCache, ackFunc)
 
-	if len(s.Buffer) >= int(s.Config.BufferSize) {
-		bufferedRecords := s.Buffer
-		s.Buffer = make([]record.Record, 0, s.Config.BufferSize)
+	if len(d.Buffer) >= int(d.Config.BufferSize) {
+		bufferedRecords := d.Buffer
+		d.Buffer = make([]sdk.Record, 0, d.Config.BufferSize)
 
-		err := s.Writer.Write(ctx, &writer.Batch{
+		// write batch into S3
+		err := d.Writer.Write(ctx, &writer.Batch{
 			Records: bufferedRecords,
-			Format:  s.Config.Format,
+			Format:  d.Config.Format,
 		})
-
 		if err != nil {
-			s.Error = err
+			d.Error = err
+		}
+
+		// call all the written records' ackFunctions
+		for _, ack := range d.AckFuncCache {
+			err := ack(d.Error)
+			if err != nil {
+				return err
+			}
 		}
+		// clear ackFunc cache
+		d.AckFuncCache = make([]sdk.AckFunc, 0, d.Config.BufferSize)
```

Comment:
Sidenote: we could make this more performant by not allocating a new slice but rather reusing the old one (same goes for `d.Buffer`.
```suggestion
		d.AckFuncCache = d.AckFuncCache[:0]
```